nexus_constructor/instrument_view/q_components.py

```diff
@@ -0,0 +1,168 @@
+from typing import List
+
+import numpy as np
+from PySide2.Qt3DCore import Qt3DCore
+from PySide2.Qt3DExtras import Qt3DExtras
+from PySide2.Qt3DRender import Qt3DRender
+from PySide2.QtGui import QColor, QMatrix4x4, QVector3D
+
+from nexus_constructor.instrument_view.off_renderer import OffMesh
+from nexus_constructor.instrument_view.qentity_utils import (
+    MATERIAL_ALPHA,
+    MATERIAL_COLORS,
+    MATERIAL_DIFFUSE_COLORS,
+    create_material,
+    create_qentity,
+)
+
+
+class QComponent:
+    def __init__(self, root_entity: Qt3DCore.QEntity, nx_class: str):
+        self.root_entity = root_entity
+        self.nx_class = nx_class
+
+        self.entities: List[Qt3DCore.QEntity] = []
+        self.default_material: Qt3DRender.QMaterial = self._create_default_material()
+
+    def create_entities(self):
+        raise NotImplementedError
+
+    def add_transformation(self, transformation: Qt3DCore.QComponent):
+        raise NotImplementedError
+
+    def remove_transformation(self, transformation: Qt3DCore.QComponent):
+        for entity in self.entities:
+            entity.removeComponent(transformation)
+
+    def setParent(self, value=None):
+        for entity in self.entities:
+            entity.setParent(value)
+
+    def get_entity(self):
+        return self.entities[0]
+
+    def _create_default_material(self) -> Qt3DRender.QMaterial:
+        return create_material(
+            MATERIAL_COLORS.get(self.nx_class, QColor("black")),
+            MATERIAL_DIFFUSE_COLORS.get(self.nx_class, QColor("grey")),
+            self.root_entity,
+            MATERIAL_ALPHA.get(self.nx_class),
+        )
+
+
+class OffMeshQComponent(QComponent):
+    def __init__(self, mesh: OffMesh, root_entity: Qt3DCore.QEntity, nx_class: str):
+        super().__init__(root_entity, nx_class)
+        self._mesh = mesh
+
+    def create_entities(self):
+        self.entities.append(
+            create_qentity([self._mesh, self.default_material], self.root_entity)
+        )
+
+    def add_transformation(self, transformation: Qt3DCore.QComponent):
+        self.entities[0].addComponent(transformation)
+
+
+class NeutronSource(QComponent):
+    def __init__(self, root_entity, nx_class):
+        super().__init__(root_entity, nx_class)
+        self._source_length = 4
+        self._source_radius = 1
+        self._num_neutrons = 8
+        self._neutron_offsets = self._generate_random_points_in_cylinder(
+            self._num_neutrons, self._source_radius, self._source_length
+        )
+
+    def create_entities(self):
+        self._create_source()
+        self._setup_neutrons()
+
+    def add_transformation(self, transformation: Qt3DCore.QComponent):
```

Comment:
I think this function can be simpler. One suggestion, why not to the entities list append tuples where first element is the entity itself and the second element is either _get_cylinder_transformatrion_matrix() or self._get_sphere_transformation_matrix(offset) and then you do not need to protected methods to set the transformations. Also you can get rid of the "if index" thing etc. Does it make sense?